### PR TITLE
Record and replay traffic from tests against internet DNS servers

### PR DESCRIPTION
**Pull request by divergentdave**
This addresses #2279 by adding a test utility that can save and replay transactions with internet-based DNS servers. If the environment variable `DNS_RECORDER` is set to `record`, it will make a connection to the remote server, record everything it proxies, and save it to a JSON file. By default, if the environment variable is not set, it will read from that file, and replay responses to matching queries. If `text2pcap` and `tshark` is present, they will be used to dissect the messages, and the dissection is saved alongside the raw bytes, to help explain changes when a test is re-recorded.

### CODE REVIEW

File: crates/dns-vcr/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+rust-version.workspace = true
+homepage.workspace = true
+repository.workspace = true
+keywords.workspace = true
+categories.workspace = true
+license.workspace = true
+publish = false
+
+[dependencies]
+hex = { version = "0.4.3", features = ["serde"] }
```
Comment (djc) [inline crates/dns-vcr/Cargo.toml:15]: Should use `hex` via the workspace dependencies too.

File: crates/dns-vcr/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+[package]
+name = "dns-vcr"
```
Comment (djc) [inline crates/dns-vcr/Cargo.toml:2]: Not a big fan of the name? Would suggest something more direct like `dns-recorder` or `dns2json` which doesn't rely on a decade-old device metaphor acronym.

File: crates/dns-vcr/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+    path: PathBuf,
+    inner: Arc<Mutex<VcrInner>>,
+    handle: JoinHandle<()>,
+    stop: Arc<AtomicBool>,
+}
+
+struct VcrInner {
+    source: VcrSource,
+    error: Option<Error>,
+}
```
Comment (djc) [inline crates/dns-vcr/src/lib.rs:102]: Nit: keep `struct UdpDnsVcr` and `impl UdpDnsVcr` together, please.

File: crates/dns-vcr/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+    thread::{self, JoinHandle},
+    time::{Duration, Instant},
+};
+
+use serde::{Deserialize, Serialize};
+use tracing::{debug, error};
+
+mod wireshark;
+
+/// Receive timeout used when listening on the mock server.
```
Comment (djc) [inline crates/dns-vcr/src/lib.rs:23]: Nit: prefer a top-down topological ordering, high-level public API first, then have these `const`s at the bottom (before `mod tests`). `Error` should thus also be closer to the bottom.

File: crates/dns-vcr/src/wireshark.rs [hunk 5]
```diff
@@ -0,0 +1,79 @@
+//! Optional support for dissecting DNS messages using Wireshark.
+
+use std::{
```
Comment (djc) [inline crates/dns-vcr/src/wireshark.rs:3]: At 75 lines, suggest just inlining this into `lib.rs`.

**Review by djc — COMMENTED** [review]
Seems like a useful tool! Some early feedback.

**Comment by divergentdave** [comment]
I have added TCP support in addition to UDP support, and instrumented a number of unit tests. I left DoT, DoQ, and DoH for future work, as well as uses of `NameServerPool`/`NameServerConfigGroup` that talk to multiple internet-based servers at once. With this PR, about a third of the tests that depended on internet access no longer do.

**Comment by divergentdave** [comment]
I tried running this under `datefudge`, about a month in the future, and a number of tests failed, either because the set of query messages changed or the DNSSEC validator returned "bogus". I think this strategy will only work if we mock the current time to match the time the recording was made. Otherwise, we'll have to replace internet-based tests with tests in `e2e-tests`.

**Pull request closed without merging**